test/replication-luatest/gh_9399_raft_leader_steps_off_on_ER_WAL_IO_test.lua

```diff
@@ -0,0 +1,92 @@
+local cluster = require('luatest.replica_set')
+local server = require('luatest.server')
+
+local g = t.group('raft-leader-steps-off-on-ER_WAL_IO')
+
+local function wait_pair_sync(server1, server2)
+    -- Without retrying it fails sometimes when vclocks are empty and both
+    -- instances are in 'connect' state instead of 'follow'.
+    t.helpers.retrying({timeout = 10}, function()
+        server1:wait_for_vclock_of(server2)
+        server2:wait_for_vclock_of(server1)
+        server1:assert_follows_upstream(server2:get_instance_id())
+        server2:assert_follows_upstream(server1:get_instance_id())
+    end)
+end
+
+g.before_each(function(cg)
+    t.tarantool.skip_if_not_debug()
+
+    cg.cluster = cluster:new({})
+
+    local box_cfg = {
+        replication = {
+            server.build_listen_uri('master', cg.cluster.id),
+            server.build_listen_uri('replica', cg.cluster.id),
+        },
+        election_mode = 'candidate',
+        replication_synchro_quorum = 1,
+        replication_synchro_timeout = 100000,
+        replication_timeout = 0.1,
+        election_fencing_mode = 'off',
+    }
+    cg.master = cg.cluster:build_and_add_server({
+        alias = 'master',
+        box_cfg = box_cfg
+    })
+    box_cfg.election_mode = 'off'
+    cg.replica = cg.cluster:build_and_add_server({
+        alias = 'replica',
+        box_cfg = box_cfg
+    })
+    cg.cluster:start()
+    cg.master:wait_until_election_leader_found()
+    cg.replica:wait_until_election_leader_found()
+
+    cg.master:exec(function()
+        box.schema.space.create('test', {is_sync = true})
+        box.space.test:create_index('pk')
+    end)
+    wait_pair_sync(cg.replica, cg.master)
+end)
+
+g.after_each(function(cg)
+    cg.cluster:drop()
+end)
+
+g.test_cluster_elect_new_leader = function(cg)
+    cg.replica:exec(function()
+        box.cfg{election_mode='candidate'}
+    end)
+    cg.master:exec(function()
+        t.assert_equals(box.info.election.state, 'leader')
+        box.error.injection.set('ERRINJ_WAL_IO', true)
+        local ok, err = pcall(box.space.test.insert, box.space.test, {1})
+        t.assert(not ok)
+        t.assert_equals(err.type, 'ClientError')
+        t.assert_equals(err.message, 'Failed to write to disk')
+    end)
+    cg.replica:wait_for_election_state('leader')
+    -- Assert that master failed on writing current raft state
+    local ok, _ = pcall(cg.master.wait_for_vclock_of, cg.master, cg.replica)
+    t.assert(not ok)
+end
+
+-- Сheck that after one unsuccessful attempt to write the leader does not fails,
+-- but a new election round starts
+g.test_leader_continues_work_after_one_write_error = function(cg)
+    local term = cg.master:get_election_term()
+    cg.master:exec(function()
+        t.assert_equals(box.info.election.state, 'leader')
+        box.error.injection.set('ERRINJ_WAL_IO', true)
+        -- Only one write attempt fails
+        box.error.injection.set('ERRINJ_WAL_IO_COUNTDOWN', 0)
```

Comment:
Why do you need a countdown error for this? Can you simply unset `ERRINJ_WAL_IO` manually back to false after the first error?

Reply:
I may not have time to unset `ERRINJ_WAL_IO` manually back to false before the node tries to write a term and crashes. 
https://github.com/tarantool/tarantool/blob/ebb170cb8cf2b9c4634bcf0178665909f578c335/src/box/raft.c#L438-L442
And I have no way to stop the node before it tries to write. `ERROR_INJECT(ERRINJ_WAL_IO, ...` is located earlier than `ERROR_INJECT_SLEEP(ERRINJ_WAL_DELAY)`.
I don't know why `ERROR_INJECT(ERRINJ_WAL_IO, ...` is located in `wal_write_async`. It might be worth moving `ERROR_INJECT(ERRINJ_WAL_IO, ...` to `wal_write_to_disk`.
Do something like there:
```patch
 	ERROR_INJECT_SLEEP(ERRINJ_WAL_DELAY);

	ERROR_INJECT_COUNTDOWN(ERRINJ_WAL_DELAY_COUNTDOWN, {
		struct errinj *e = errinj(ERRINJ_WAL_DELAY, ERRINJ_BOOL);
		e->bparam = true;
		ERROR_INJECT_SLEEP(ERRINJ_WAL_DELAY);
	});

+	ERROR_INJECT(ERRINJ_WAL_IO, {
+		err_code = JOURNAL_ENTRY_ERR_IO;
+		goto done;
+	}

	if (writer->is_in_rollback) {
		/* We're rolling back a failed write. */
		err_code = JOURNAL_ENTRY_ERR_CASCADE;
		goto done;
	}

	/* Xlog is only rotated between queue processing  */
	if (wal_opt_rotate(writer) != 0) {
		err_code = JOURNAL_ENTRY_ERR_IO;
		goto done;
	}
```

Reply:
Done.